lib/sycamore/sycamore/transforms/extract_graph.py

```diff
@@ -0,0 +1,155 @@
+            value = str(doc["properties"][m.nodeKey])
+            node = {"type": "metadata", "properties": {key: value}, "label": str(m.nodeLabel), "relationships": {}}
+            nodes[key + "_" + value] = node
+
+            rel = {
+                "TYPE": str(m.relLabel),
+                "properties": {},
+                "START_ID": str(doc.doc_id),
+                "START_LABEL": doc.data["label"],
+            }
+            nodes[key + "_" + value]["relationships"][str(uuid.uuid4())] = rel
+            del doc["properties"][m.nodeKey]
+
+        doc["properties"]["nodes"] = nodes
+        return doc
+
+    def resolve_metadata(self, docset: "DocSet") -> "DocSet":
+        """
+        Aggregates 'nodes' from every document and resolves duplicate nodes
+        """
+        from sycamore import Execution
+        from sycamore.reader import DocSetReader
+        from ray.data.aggregate import AggregateFn
+
+        execution = Execution(docset.context, docset.plan)
+        dataset = execution.execute(docset.plan)
+
+        def extract_nodes(row):
+            doc = Document.deserialize(row["doc"])
+            if isinstance(doc, MetadataDocument) or "nodes" not in doc["properties"]:
+                return {}
+            return doc["properties"]["nodes"]
+
+        def accumulate_row(nodes, row):
+            extracted = extract_nodes(row)
+            for key, value in extracted.items():
+                if nodes.get(key, {}) == {}:
+                    nodes[key] = value
+                else:
+                    for rel_uuid, rel in extracted[key]["relationships"].items():
+                        nodes[key]["relationships"][rel_uuid] = rel
+            return nodes
+
+        def merge(nodes1, nodes2):
+            for key, value in nodes2.items():
+                if nodes1.get(key, {}) == {}:
+                    nodes1[key] = value
+                else:
+                    for rel_uuid, rel in nodes2[key]["relationships"].items():
+                        nodes1[key]["relationships"][rel_uuid] = rel
+            return nodes1
+
+        def finalize(nodes):
+            for value in nodes.values():
+                value["doc_id"] = str(uuid.uuid4())
+                for rel in value["relationships"].values():
+                    rel["END_ID"] = value["doc_id"]
+                    rel["END_LABEL"] = value["label"]
+            return nodes
+
+        aggregation = AggregateFn(
+            init=lambda group_key: {}, accumulate_row=accumulate_row, merge=merge, finalize=finalize, name="nodes"
+        )
+
+        result = dataset.aggregate(aggregation)
+        docs = dataset.take_all(None)
+        docs = [Document.deserialize(d["doc"]) for d in docs]
+
+        for doc in docs:
+            if "properties" in doc:
+                if "nodes" in doc["properties"]:
+                    del doc["properties"]["nodes"]
+
+        for value in result["nodes"].values():
+            docs[0]["elements"].append(value)
```

Comment:
Yeah so this is because once we explode after this step, all the elements become top level documents(so in this case each node gets turned into a top level document). Then I feed this into a neo4j loader. I wish there was a better solution cause this seems scrappy. do you have any ideas?